src/Nerdbank.Streams/Sequence`1.cs

```diff
@@ -0,0 +1,258 @@
+    /// </summary>
+    /// <typeparam name="T">The type of element stored by the sequence.</typeparam>
+    /// <remarks>
+    /// Instance members are not thread-safe.
+    /// </remarks>
+    public class Sequence<T>
+    {
+        private readonly Stack<SequenceSegment> segmentPool = new Stack<SequenceSegment>();
+
+        private readonly MemoryPool<T> memoryPool;
+
+        private SequenceSegment first;
+
+        private SequenceSegment last;
+
+        public Sequence()
+            : this(MemoryPool<T>.Shared)
+        {
+        }
+
+        public Sequence(MemoryPool<T> memoryPool)
+        {
+            Requires.NotNull(memoryPool, nameof(memoryPool));
+            this.memoryPool = memoryPool;
+        }
+
+        public static implicit operator ReadOnlySequence<T>(Sequence<T> sequence)
+        {
+            return sequence.first != null
+                ? new ReadOnlySequence<T>(sequence.first, 0, sequence.last, sequence.last.Length)
+                : ReadOnlySequence<T>.Empty;
+        }
+
+        public void Append(IMemoryOwner<T> array) => this.Append(array, 0, Requires.NotNull(array, nameof(array)).Memory.Length);
+
+        public void Append(IMemoryOwner<T> array, int length) => this.Append(array, 0, length);
+
+        public void Append(IMemoryOwner<T> array, int start, int length)
+        {
+            Requires.NotNull(array, nameof(array));
+            Requires.Range(length >= 0 && start + length <= array.Memory.Length, nameof(length));
+
+            var segment = this.segmentPool.Count > 0 ? this.segmentPool.Pop() : new SequenceSegment();
+            segment.SetMemory(array, start, start + length);
+            if (this.last == null)
+            {
+                this.first = this.last = segment;
+            }
+            else
+            {
+                this.last.SetNext(segment);
+                this.last = segment;
+            }
+        }
+
+        public void Advance(SequencePosition position)
+        {
+            // TODO: protect against SequencePosition arguments that do not represent a FORWARD position for THIS sequence.
+
+            var firstSegment = (SequenceSegment)position.GetObject();
+            int firstIndex = position.GetInteger();
+
+            var oldFirst = this.first;
+            while (oldFirst != firstSegment)
+            {
+                oldFirst.ResetMemory();
+                oldFirst = oldFirst.Next;
+            }
+
+            firstSegment.AdvanceTo(firstIndex);
+
+            if (firstSegment.Length == 0)
+            {
+                firstSegment = this.RecycleAndGetNext(firstSegment);
+            }
+
+            this.first = firstSegment;
+
+            if (this.first == null)
+            {
+                this.last = null;
+            }
+        }
+
+        // TODO: change return type to Memory<T>
+        public IMemoryOwner<T> GetMemory(int minimumLength)
```

Comment:
I'm not a fan of the fact that GetMemory returns a detached piece of memory that needs to be appended. Is that by design? The Pipe doesn't allow this today. Can you think of a scenario where that's useful? 

I think you either return an attached segment or you let the caller allocate and Append becomes the only API.

Reply:
Thanks. That's very good feedback. The reason I did this originally is since I want this class to automatically return arrays that are no longer used to a pool, and it should always be the *same* pool from which the array was obtained, it made sense to me to help ensure that by providing a method to obtain the memory.

It sounds like you'd be OK with this so long as it returns "attached" memory. What do you mean by that? Is it that I consider it part of the sequence? If I remove `Append`, I would need to replace it with a `PipeWriter.Advance`-like method. I already have a `PipeReader.AdvanceTo`-like method. While I'd love for my methods to resemble the new Pipe method API since I imagine they'll both be used by the same developers, having `Advance` and `AdvanceTo` on the same class would probably be confusing. This isn't a problem for Pipes since the methods are on different classes.

Suggestions?

Reply:
For now I'm using `Advance` and `AdvanceTo`, per the `PipeReader` and `PipeWriter` precedent. I may rename the methods for clarity as they appear on one type, but I think it's getting to what you were asking for, in that we no longer have an Append method.